### PR TITLE
[WebAssembly] Don't return multivalue when Emscripten EH/SjLj is used

### PR DESCRIPTION
**Pull request by aheejin**
Emscripten EH/SjLj uses invoke wrappers in JS, such as `invoke_vi`, from which user functions are called indirectly, to emulate exceptions and setjmp/longjmp.

But in case the invoked function returns multiple values or a 128-bit value, its the JS invoke wrappers cannot return multivalue because JS doesn't support that. So we should not enable multivalue returns for the JS invoke wrappers and also the functions called by those JS wrappers because their signature has to match with the JS wrapper. For example, if `func` returns `{i32, i32}` and we have
```ll
  invoke {i32, i32} @func() ...
```
while LowerEmscriptenEHSjLj will lower it down to something like
```ll
  %0 = call { i32, i32 } @"__invoke_{i32.i32}_void"(ptr @func)
  ...
```
we should eventually lower both the invoke wrapper (whose name will be changed later to `invoke_vi`) and `func` down to a signature that indirectly returns multiple values by memory parameter, because JS invoke wrappers do support multiple return values.

So we need to disable multivalue returns for JS invoke wrappers and functions called by them. I think we have three ways to do that:

1. Make a set and add all functions that are invoked by JS invoke  wrappers in LowerEmscriptenEHSjLj and pass it to the backend using an auxiliary data structure. We have a precedent of this kind of structure already, which is used for Wasm EH:
   https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/CodeGen/WasmEHFuncInfo.h
   We can even add this set to `WasmEHFuncInfo` maybe, given that this is also a way of handling exceptions in Wasm.
   - Pros: Most precise
   - Cons: Auxiliary structure needed

2. Unless we record the invoked functions in LowerEmscriptenEHSjLj like 1, we don't have a way of precisely knowing the set of invoked functions. But they are all indirectly invoked, so in the backend we can check whether a given function is ever indirectly used (i.e., its pointer taken) by traversing its `users()` in the IR and if it is, we don't allow multivalue returns for them.
   - Pros: No auxiliary structure
   - Cons: IR checking overhead. More conservative than 1.

3. Disallow all multivalue returns when Emscripten EH or SjLj is enabled.
   - Pros: Simplest
   - Cons: Most conservative.

This PR is doing 3. While it is the most conservative and possibly disallowing multivalue returns from more functions than needed, I chose this because it is the simplest, and given that hopefully more people will adopt Wasm EH going forward, I don't think there will be many people who would use multivalue and Emscripten EH/SjLj together and want the whatever performance benefit that multivalue return can bring, given that Emscripten EH/SjLj has already a huge performance cost.

This is separate from whether we should make the multivalue return dependent on the multivalue feature or something else like a clang flag, which is being partly discussed in https://github.com/WebAssembly/tool-conventions/pull/223.
Whichever way we decide on that front, we still need to disable multivalue returns in case Emscripten EH/SjLj is used.

### CODE REVIEW

**Comment by aheejin** [comment]
cc @TerrorJack in case you are interested. If you are using emscripten this may affect your results, but I don't think you are..?

**Comment by tlively** [comment]
> the JS invoke wrappers cannot return multivalue because JS doesn't support that

Doesn't it? I thought multivalue interacts with JS by having multivalue function called from JS return JS arrays of values.

File: llvm/lib/Target/WebAssembly/WebAssemblyMachineFunctionInfo.cpp [hunk 1]
```diff
@@ -70,12 +71,12 @@ void llvm::computeSignatureVTs(const FunctionType *Ty,
...
-      !TM.getSubtarget<WebAssemblySubtarget>(ContextFunc).hasMultivalue()) {
+  if (!WebAssembly::canLowerReturn(
+          Results.size(),
+          &TM.getSubtarget<WebAssemblySubtarget>(ContextFunc))) {
     // WebAssembly can't lower returns of multiple values without demoting to
-    // sret unless multivalue is enabled (see
-    // WebAssemblyTargetLowering::CanLowerReturn). So replace multiple return
-    // values with a poitner parameter.
+    // sret unless multivalue is enabled (see WebAssembly::canLowerReturn). So
+    // replace multiple return values with a poitner parameter.
```
Comment (tlively) [inline llvm/lib/Target/WebAssembly/WebAssemblyMachineFunctionInfo.cpp:79]:
```suggestion
    // replace multiple return values with a pointer parameter.
```

File: llvm/test/CodeGen/WebAssembly/lower-em-ehsjlj-multi-return.ll [hunk 2]
```diff
@@ -1,8 +1,8 @@
-; RUN: not --crash llc < %s -enable-emscripten-cxx-exceptions -mattr=+multivalue 2>&1 | FileCheck %s --check-prefix=EH
-; RUN: not --crash llc < %s -enable-emscripten-sjlj -mattr=+multivalue 2>&1 | FileCheck %s --check-prefix=SJLJ
+; RUN: llc < %s -enable-emscripten-cxx-exceptions -enable-emscripten-sjlj -mattr=+multivalue | FileCheck %s
```
Comment (tlively) [inline llvm/test/CodeGen/WebAssembly/lower-em-ehsjlj-multi-return.ll:1]: Looks like there are some stale check lines at the end of the file.

**Comment by tlively** [comment]
> One thing to think about is, given that Emscripten EH is off by default but Emscripten SjLj is on by default unless you use Wasm EH, this may turn off multivalue returns for more cases than we intend to. So to sum up,
> 
> * Enabling Emscripten EH automatically means enabling Emscripten SjLj
> * Enabling Wasm EH automatically means enabling Wasm SjLj
> * Enabling neither EH means enabling Emscripten SjLJ
>   
>   * This can change later when Wasm EH is supported universally, but likely not in the near future

So if you really want multivalue returns, the only way to get them after this change is by also enabling Wasm EH?

**Comment by TerrorJack** [comment]
@aheejin Thanks for the ping! Our toolchain does not use emscripten sjlj mechanism fyi.

**Comment by aheejin** [comment]
> > the JS invoke wrappers cannot return multivalue because JS doesn't support that
> 
> Doesn't it? I thought multivalue interacts with JS by having multivalue function called from JS return JS arrays of values.

Hmm, yeah, you're right. It's not an inherent limitation of JS but rather our invoke wrapper-specific issue...

**Comment by aheejin** [comment]
> > One thing to think about is, given that Emscripten EH is off by default but Emscripten SjLj is on by default unless you use Wasm EH, this may turn off multivalue returns for more cases than we intend to. So to sum up,
> > 
> > * Enabling Emscripten EH automatically means enabling Emscripten SjLj
> > * Enabling Wasm EH automatically means enabling Wasm SjLj
> > * Enabling neither EH means enabling Emscripten SjLJ
> >   
> >   * This can change later when Wasm EH is supported universally, but likely not in the near future
> 
> So if you really want multivalue returns, the only way to get them after this change is by also enabling Wasm EH?

Either that or explicitly changing longjmp support via `-sSUPPORT_LONGJMP=wasm` or `-sSUPPORT_LONGJMP=0`. But yeah I don't think people would do that. I was mostly thinking about Emscripten EH when I was working on this and disabling multivalue whenever Emscripten SjLj is enabled can be too much. I guess I need to rethink; will make the PR a draft in the meantime.

File: llvm/test/CodeGen/WebAssembly/lower-em-ehsjlj-multi-return.ll (hunk 2 above)
Comment (aheejin) [reply llvm/test/CodeGen/WebAssembly/lower-em-ehsjlj-multi-return.ll:1]: Removed.

**Comment by aheejin** [comment]
Will abandon this given that Emscripten does not need this for now after #88492 lands.

**Pull request closed without merging**